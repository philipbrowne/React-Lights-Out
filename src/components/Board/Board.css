.Board {
  text-align: center;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-color: rgba(0, 0, 0, 0.1);
}

.Board-Table {
  width: 80vmin;
  height: 80vmin;
}

.Board-Header {
  color: orange;
  margin-bottom: 1rem;
  font-size: 2rem;
}
