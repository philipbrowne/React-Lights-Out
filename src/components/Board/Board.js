import React, { useState } from 'react';
import Cell from '../Cell/Cell';
import GameRules from '../GameRules/GameRules';
import './Board.css';

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

const isLit = (odds) => {
  const num = Math.random();
  return num < odds;
};

function Board({ nrows = 5, ncols = 5, chanceLightStartsOn = 0.5 }) {
  const [board, setBoard] = useState(createBoard());

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {
    let initialBoard = Array.from({ length: nrows });
    for (let y = 0; y < initialBoard.length; y++) {
      initialBoard[y] = Array.from({ length: ncols });
    }
    for (let y = 0; y < nrows; y++) {
      for (let x = 0; x < ncols; x++) {
        initialBoard[y][x] = isLit(chanceLightStartsOn);
      }
    }
    return initialBoard;
  }

  function hasWon(board) {
    for (let y = 0; y < nrows; y++) {
      for (let x = 0; x < ncols; x++) {
        if (board[y][x] === true) {
          return false;
        }
      }
    }
    return true;
  }

  function flipCellsAround(coord) {
    setBoard((oldBoard) => {
      const [y, x] = coord.split('-').map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it

        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };
      let boardCopy = [...oldBoard];
      flipCell(y, x, boardCopy);
      if (x > 0) {
        flipCell(y, x - 1, boardCopy);
      }
      if (x < ncols) {
        flipCell(y, x + 1, boardCopy);
      }
      if (y > 0) {
        flipCell(y - 1, x, boardCopy);
      }
      if (y < ncols) {
        flipCell(y + 1, x, boardCopy);
      }
      return boardCopy;
    });
  }

  let gameBoard = [];
  for (let y = 0; y < nrows; y++) {
    let row = [];
    for (let x = 0; x < ncols; x++) {
      let coord = `${y}-${x}`;
      row.push(
        <Cell
          ncols={ncols}
          nrows={nrows}
          key={coord}
          isLit={board[y][x]}
          flipCellsAroundMe={() => flipCellsAround(coord)}
        />
      );
    }
    gameBoard.push(<tr key={y}>{row}</tr>);
  }
  if (hasWon(board)) {
    return (
      <h1 className="Board-Header" data-testid="winning-header">
        Congratulations You Won!
      </h1>
    );
  }
  return (
    <div className="Board">
      <h1 className="Board-Header">Lights Out!</h1>
      <table className="Board-Table">
        <tbody>{gameBoard}</tbody>
      </table>
      <GameRules />
    </div>
  );
}

export default Board;
