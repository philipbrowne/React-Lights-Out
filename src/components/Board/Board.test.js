import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Board from './Board';

describe('Board tests', () => {
  it('renders without crashing', () => {
    render(<Board />);
  });
  it('matches snapshot for board with all lights on', () => {
    const { asFragment } = render(<Board chanceLightStartsOn={1} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('matches snapshot for board with all lights off', () => {
    const { asFragment } = render(<Board chanceLightStartsOn={0} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('handles cell clicks properly', () => {
    const { getAllByRole } = render(
      <Board nrows={3} ncols={3} chanceLightStartsOn={1} />
    );
    const cells = getAllByRole('button');
    cells.forEach((cell) => {
      expect(cell).toHaveClass('Cell-lit');
    });

    //Click Middle Cell
    fireEvent.click(cells[4]);
    // Make sure clicked cell is now no longer lit
    expect(cells[4]).not.toHaveClass('Cell-lit');

    //Cells vert and horiz adjacent to middle cell
    let adjCells = [cells[1], cells[3], cells[5], cells[7]];
    // Make sure vert and horiz adjacent cells are now no longer lit
    adjCells.forEach((cell) => {
      expect(cell).not.toHaveClass('Cell-lit');
    });

    //Corner cells not adjacent to middle cell
    let cornerCells = [cells[0], cells[2], cells[6], cells[8]];
    // Make sure corner cells are still lit
    cornerCells.forEach((cell) => {
      expect(cell).toHaveClass('Cell-lit');
    });
  });
  it('handles a winning condition properly', () => {
    const { getAllByRole, queryByTestId } = render(
      <Board nrows={1} ncols={3} chanceLightStartsOn={1} />
    );
    const cells = getAllByRole('button');
    expect(queryByTestId('winning-header')).not.toBeInTheDocument();
    fireEvent.click(cells[1]);
    expect(queryByTestId('winning-header')).toBeInTheDocument();
  });
});
