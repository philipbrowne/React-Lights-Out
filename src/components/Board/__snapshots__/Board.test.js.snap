// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board tests matches snapshot for board with all lights off 1`] = `
<DocumentFragment>
  <h1
    class="Board-Header"
    data-testid="winning-header"
  >
    Congratulations You Won!
  </h1>
</DocumentFragment>
`;

exports[`Board tests matches snapshot for board with all lights on 1`] = `
<DocumentFragment>
  <div
    class="Board"
  >
    <h1
      class="Board-Header"
    >
      Lights Out!
    </h1>
    <table
      class="Board-Table"
    >
      <tbody>
        <tr>
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
          <td
            class="Cell Cell-lit"
            role="button"
            style="width: 20%; height: 20%;"
          />
        </tr>
      </tbody>
    </table>
    <div
      class="GameRules"
    >
      <p>
        The game consists of a grid of lights which is 5x5 by default. When the game starts, a random number of these lights is switched on. The game default is a 50% chance of being lit. Pressing any of the lights will toggle it and the four adjacent lights. The goal of the puzzle is to switch all the lights off, preferably in as few button presses as possible.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
